services/vm-watcher/vm_watcher_service.go

```diff
@@ -0,0 +1,220 @@
+		ctx:      ctx,
+		provider: provider,
+	}
+}
+
+var _ manager.LeaderElectionRunnable = Service{}
+
+func (s Service) NeedLeaderElection() bool {
+	return true
+}
+
+func (s Service) Start(ctx context.Context) error {
+	ctx = cource.JoinContext(ctx, s.ctx)
+	ctx = watcher.JoinContext(ctx, s.ctx)
+	ctx = pkgcfg.JoinContext(ctx, s.ctx)
+
+	ctx = logr.NewContext(
+		ctx,
+		logr.FromContextOrDiscard(s.ctx).WithName("VMWatcherService"))
+
+	for ctx.Err() == nil {
+		if err := s.waitForChanges(ctx); err != nil {
+			// If waitForChanges failed because of an invalid login or auth
+			// error, then do not treat the error as fatal. This allows the
+			// loop to run again, kicking off another watcher with what should
+			// be updated credentials.
+			if !vsphereclient.IsInvalidLogin(err) &&
+				!vsphereclient.IsNotAuthenticatedError(err) {
+
+				return err
+			}
+		}
+	}
+	return ctx.Err()
+}
+
+var emptyResult watcher.Result
+
+func (s Service) vmFolderRefs(ctx context.Context) ([]vimtypes.ManagedObjectReference, error) {
+	// Get a list of all the folders that can contain VM Service VMs.
+	var (
+		zones topologyv1.ZoneList
+		frefs []vimtypes.ManagedObjectReference
+		moids = map[string]struct{}{}
+	)
+	if err := s.Client.List(ctx, &zones); err != nil {
+		return nil, err
+	}
+	for i := range zones.Items {
+		z := zones.Items[i]
+		if v := z.Spec.ManagedVMs.FolderMoID; v != "" {
+			if _, ok := moids[v]; !ok {
+				moids[v] = struct{}{}
+				frefs = append(frefs, vimtypes.ManagedObjectReference{
+					Type:  "Folder",
+					Value: v,
+				})
+			}
+		}
+	}
+
+	return frefs, nil
+}
+
+func (s Service) waitForChanges(ctx context.Context) error {
+	var (
+		logger     = logr.FromContextOrDiscard(ctx)
+		chanSource = cource.FromContext(ctx, "VirtualMachine")
+	)
+
+	vcClient, err := s.provider.VSphereClient(ctx)
+	if err != nil {
+		return err
+	}
+	logger.Info("got vsphere client")
+
+	refs, err := s.vmFolderRefs(ctx)
+	if err != nil {
+		return err
+	}
+	logger.Info("got vm service folders", "refs", refs)
+
+	// Start the watcher.
+	w, err := watcher.Start(
+		ctx,
+		vcClient.VimClient(),
+		nil,
+		nil,
+		s.lookupNamespacedName,
+		refs...)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("started watching vms")
+
+	for {
+		select {
+		case result := <-w.Result():
+			if result == emptyResult {
+				logger.Info("received empty result, watcher is closed")
+				return w.Err()
+			}
+
+			if !result.Verified {
+				// Validate the vm exists on this cluster.
+				result.Verified = s.Get(
+					ctx,
+					ctrlclient.ObjectKey{
+						Namespace: result.Namespace,
+						Name:      result.Name,
+					},
+					&vmopv1.VirtualMachine{}) == nil
+			}
+
+			if !result.Verified {
+				logger.V(4).Info(
+					"received result but unable to validate vm",
+					"result", result)
+				continue
+			}
+
+			// Enqueue a reconcile request for the VM.
+			logger.V(4).Info("received result", "result", result)
+			go func(result watcher.Result) {
```

Comment:
Do we lose any useful backpressure on by doing this in a go routine? I guess I'm a little worried we could have these stackup.

Reply:
Stacking is fine. It's the same as a work queue.